pkg/api/utils.go

```diff
@@ -104,3 +111,101 @@ func handleMetricData(currentTask *db.TaskModel, updatedTask *db.TaskModel) {
 		}()
 	}
 }
+
+// Get the user's IP address from the gin request headers
+func getCallerIP(c *gin.Context) string {
+	if runtimeEnv := utils.LoadDotEnv("RUNTIME_ENV"); runtimeEnv == "aws" {
+		callerIp := c.Request.Header.Get("X-Original-Forwarded-For")
+		log.Info().Msgf("Got caller IP from X-Original-Forwarded-For header: %s", callerIp)
+		return callerIp
+	}
+	return c.ClientIP()
+}
+
+// CustomGinLogger logs a gin HTTP request in format.
+// Allows to set the logger for testing purposes.
+func CustomGinLogger(logger *zerolog.Logger) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		start := time.Now() // Start timer
+		path := c.Request.URL.Path
+		raw := c.Request.URL.RawQuery
+
+		// Read the request body
+		body, err := io.ReadAll(c.Request.Body)
+		if err != nil {
+			log.Printf("Error reading request body: %v", err)
+			c.AbortWithStatus(500)
+			return
+		}
+
+		// Log the size of the request body
+		requestSize := len(body)
+		log.Printf("Request size: %d bytes", requestSize)
+
+		// Restore the request body to the context
+		c.Request.Body = io.NopCloser(io.NopCloser(bytes.NewBuffer(body)))
+
+		// Process request
+		c.Next()
+
+		// Fill the params
+		param := gin.LogFormatterParams{}
+
+		param.TimeStamp = time.Now() // Stop timer
+		param.Latency = param.TimeStamp.Sub(start)
+		if param.Latency > time.Minute {
+			param.Latency = param.Latency.Truncate(time.Second)
+		}
+
+		param.ClientIP = getCallerIP(c)
+		param.Method = c.Request.Method
+		param.StatusCode = c.Writer.Status()
+		// param.ErrorMessage = c.Errors.ByType(gin.ErrorTypePrivate).String()
+		param.BodySize = c.Writer.Size()
+		if raw != "" {
+			path = path + "?" + raw
+		}
+		param.Path = path
+
+		// Log using the params
+		// statusCode := c.Writer.Status()
+
+		consoleWriter := zerolog.ConsoleWriter{
```

Comment:
this is the log output
`12:27AM GIN 200 73.014542ms ::1 GET /api/v1/tasks/510639e5-e1c5-4bfb-98d0-f0356250cc61 6456 689 ip=::1 latency=73.014542ms method=GET path=/api/v1/tasks/510639e5-e1c5-4bfb-98d0-f0356250cc61 req_size=689 resp_size=6456 status_code=200
`